crates/matrix-sdk/src/send_queue.rs

```diff
@@ -746,6 +787,162 @@ impl QueueStorage {
     async fn local_echoes(&self) -> Result<Vec<QueuedEvent>, RoomSendQueueStorageError> {
         Ok(self.client()?.store().load_send_queue_events(&self.room_id).await?)
     }
+
+    /// Try to apply a single dependent event, whether it's local or remote.
+    ///
+    /// This swallows errors that would retrigger every time if we retried
+    /// applying the dependent event: invalid edit content, etc.
+    #[instrument(skip_all)]
+    async fn try_apply_single_dependent_event(
+        &self,
+        client: &Client,
+        de: DependentQueuedEvent,
+    ) -> Result<(), RoomSendQueueError> {
+        let store = client.store();
+
+        match de.kind {
+            DependentQueuedEventKind::Edit { new_content } => {
+                if let Some(event_id) = de.event_id {
+                    // Has been sent; send an edit event.
+                    let room = client
+                        .get_room(&self.room_id)
+                        .ok_or(RoomSendQueueError::RoomDisappeared)?;
+
+                    // Check the event is one we know how to edit with an edit event.
+
+                    // 1. It must be deserializable…
+                    let content = match new_content.deserialize() {
+                        Ok(c) => c,
+                        Err(err) => {
+                            warn!("unable to deserialize: {err}");
+                            return Ok(());
+                        }
+                    };
+
+                    // 2. …and a room message, at this point.
+                    let AnyMessageLikeEventContent::RoomMessage(room_message_content) = content
+                    else {
+                        warn!("trying to send an edit event for a non-room message: aborting");
+                        return Ok(());
+                    };
+
+                    // Assume no relation.
+                    let new_content: RoomMessageEventContentWithoutRelation =
+                        room_message_content.into();
+
+                    let edit_event = match room
+                        .make_edit_event(&event_id, EditedContent::RoomMessage(new_content))
+                        .await
+                    {
+                        Ok(e) => e,
+                        Err(err) => {
+                            warn!("couldn't create edited event: {err}");
+                            return Ok(());
+                        }
+                    };
+
+                    // Queue the edit event in the send queue 🧠.
+                    let serializable = SerializableEventContent::from_raw(
+                        Raw::new(&edit_event)
+                            .map_err(RoomSendQueueStorageError::JsonSerialization)?,
+                        edit_event.event_type().to_string(),
+                    );
+
+                    store
+                        .save_send_queue_event(&self.room_id, TransactionId::new(), serializable)
```

Comment:
We create a new transaction ID, presumably because the one in `DependentQueuedEvent` is the transaction ID of the parent. It would probably been helpful to call this out in the docs of the `DependentQueuedEvent` type.